module/x/gravity/keeper/invariants.go

```diff
@@ -124,3 +141,502 @@ func sumPendingIbcAutoForwards(ctx sdk.Context, k Keeper, expectedBals map[strin
 
 	return expectedBals
 }
+
+// StoreValidityInvariant checks that the currently stored objects are not corrupted and all pass ValidateBasic checks
+// Note that the returned bool should be true if there is an error, e.g. an unexpected batch was processed
+func StoreValidityInvariant(k Keeper) sdk.Invariant {
+	return func(ctx sdk.Context) (string, bool) {
+		strErr, invalid := ValidateStore(ctx, k)
+		if invalid {
+			return strErr, true
+		}
+		// Assert that batch metadata is consistent and expected
+		strErr, invalid = CheckBatches(ctx, k)
+		if invalid {
+			return strErr, true
+		}
+
+		// Assert that valsets have been updated in the expected manner
+		strErr, invalid = CheckValsets(ctx, k)
+		if invalid {
+			return strErr, true
+		}
+
+		// Assert that pending ibc auto-forwards are only outgoing
+		strErr, invalid = CheckPendingIbcAutoForwards(ctx, k)
+		if invalid {
+			return strErr, true
+		}
+
+		// SUCCESS: If execution made it here, everything passes the sanity checks
+		return "", false
+	}
+}
+
+// STORE VALIDITY HELPERS
+
+// ValidateStore checks that all values in the store can be decoded and pass a ValidateBasic check
+// Returns an error string and a boolean indicating an error if true, for use in an invariant
+// nolint: gocyclo
+func ValidateStore(ctx sdk.Context, k Keeper) (string, bool) {
+	// TODO: Check the newly added iterators with unit tests
```

Comment:
You will notice lots of new iterators and ValidateBasic checks added which were missing. I also tried to clean up implementations where possible so there was less code duplication.